### ISSUE TITLE
Startup is blocked resolving configuration for a slow remote folder

### ISSUE THREAD

**Issue by FerrariAndrea** [microsoft/vscode#92274]
Hi, in my use case i have some folders opened in the workspace, one of these is remote, i mean that one of these is on a computer in lan network with access allowed by Samba.
When i start Visual Studio Code if this lan-computer is off Visual Studio Code is blocked for some min trying to access on the lan-folder uslessly.
It would be nice to make this access to each folder on workspace in async way, for avoid the bloking.

Thanks :)

**Comment by bpasero** [comment]
@FerrariAndrea can you clarify how you experience these minutes of waiting? Maybe with screencast?

**Comment by FerrariAndrea** [comment]
The screen does not give much more detail, this is Visual Studio blocked on startup:
![image](https://user-images.githubusercontent.com/33545798/76396643-e016ba80-6379-11ea-9291-2072943c9658.png)
And this is the folder offline, after the startup of Visual Studio Code: 
![image](https://user-images.githubusercontent.com/33545798/76397023-a2fef800-637a-11ea-9040-ec7b3ee4fbeb.png)

The NodeRealRobot is a folder added with "Add Folder to Workspace" and the folder path: "\\x.x.x.x\NodeRealRobot\" where "x.x.x.x"  is the static lan ip of the pc. (I know i can use git or other system for work in local and then push on lan-computer, but as i m doing for now is very comfortable)

Example, adding new lan-folder to worksapce:
![image](https://user-images.githubusercontent.com/33545798/76398375-0f7af680-637d-11ea-9fa5-cb22fcf48ebc.png)

I can add this the waiting time not seams always the same, but in my opinion is not only correlated with Visual Studio Code, it is probable that the waiting time depends to Windows file system waiting time for resolve Samba path.

**Comment by bpasero** [comment]
@FerrariAndrea how long does vscode stay how it is shown in the first screen shot? is vscode loaded from some share or local?

**Comment by FerrariAndrea** [comment]
The vscode installation is on local. The project folder is shared. The timing depends, not less than 40 seconds if I do nothing else in the meantime, but it seems more than 1 minute if I do other things (I'm not sure of this last). I think that vscode is waiting for windows check on the remote path as explorer.exe do if you try to go on a lan resource do not exist. If i remove the remote folder from workspace or if the remote folder is just reachable vscod is up in less then 3 sec.

**Comment by bpasero** [comment]
@FerrariAndrea yeah we do wait for the folder to load, but I would not expect the entire UI to look so frozen until that happens. Is it really 40s until something like the explorer gets drawn or do you see a progress bar before that.

**Comment by FerrariAndrea** [comment]
Yes the UI is blank for all time with no status bar, but i can move it, so it is not realy blocked.
And now i have noticed another thing: after the 40sec of block, so when the vscode is ready and i can access to my reachable folders, if i click on the unreachable folder then i see the status bar and the loading gif, vscode should do this on the startup?

In read is evidenced ther the status bar and meanwhile i have access and i can edit the reachable codes:
![image](https://user-images.githubusercontent.com/33545798/76577664-617b6380-64c6-11ea-95a8-91c6faa0d05c.png)

Maybe it is not need a new features, it is just a bug?

**Comment by bpasero** [comment]
Yes, this is what I would expect right away. The progress indicator appears for as long as the explorer is loading. 

Just to clarify again, you said:

> Hi, in my use case i have some folders opened in the workspace, one of these is remote, i mean that one of these is on a computer in lan network with access allowed by Samba.

I wonder if VSCode is trying to load anything from that samba for actually starting up that could slow it down other than the folder. Does it startup fast if you open a local folder? Are you sure there is no data persisted on the network share from VSCode?

**Comment by FerrariAndrea** [comment]
I m not sure to understand your last question. What you mean with "data persisted"? 

As now i m using vscode with only 3 folders in my workspace, 2 of these are in local and other 1 is in remote (with samba access). 
I can see the async exploration (noticed in my last message) of the remote folder when i click on it after the startup.

For do a test now i have remove the remote folder form my workspace (so now i have only 2 local folders in my workspace) and vscode start up in 3s, so I think that the cause of 40s block was the remote folder for sure.

**Comment by bpasero** [comment]
@FerrariAndrea assuming that this issue also shows when you reload the window, can we start to CPU profile this to see where most time is spend? Here is how that goes:
* Help > Toggle Developer Tools
* click on the "..." on the right hand side of the panel that opened
* click "More Tools" > "JavaScript Profiler"
* click the blue Start button
* reload the window
* wait for everything to have finished loading
* save the CPU profile and attach it here

Thanks

**Comment by FerrariAndrea** [comment]
I have add again the remote folder, i try to take a profile while start up but vscode can't do this.
So i have take a profile after startup and while vscode try to access on the remote folder after my click on it.
And this is the the profile: 
[CPU-20200317T103132.zip](https://github.com/microsoft/vscode/files/4342610/CPU-20200317T103132.zip)

I did not know ther was the console like that of Chrome, and thanks to this i can see the console.log error of vscode at startup:

> bootstrap-window.js:213 [uncaught exception]: Error: Unable to read file '\\REMOTE_FOLDER_LAN_IP\NodeRealRobot\.vscode\settings.json' (Unknown (FileSystemError): Error: UNKNOWN: unknown error, open '\\REMOTE_FOLDER_LAN_IP\NodeRealRobot\.vscode\settings.json')
> 
> Error: Unable to read file '\\REMOTE_FOLDER_LAN_IP\NodeRealRobot\.vscode\settings.json' (Unknown (FileSystemError): Error: UNKNOWN: unknown error, open '\\REMOTE_FOLDER_LAN_IP\NodeRealRobot\.vscode\settings.json')
>     at _.doReadAsFileStream (file:///C:/Users/USER_NAME/AppData/Local/Programs/Microsoft VS Code/resources/app/out/vs/workbench/workbench.desktop.main.js:1667:417)
> onUnexpectedError @ bootstrap-window.js:213
> bootstrap-window.js:216 Error: Unable to read file '\\REMOTE_FOLDER_LAN_IP\NodeRealRobot\.vscode\settings.json' (Unknown (FileSystemError): Error: UNKNOWN: unknown error, open '\\REMOTE_FOLDER_LAN_IP\NodeRealRobot\.vscode\settings.json')
> 
> Error: Unable to read file '\\REMOTE_FOLDER_LAN_IP\NodeRealRobot\.vscode\settings.json' (Unknown (FileSystemError): Error: UNKNOWN: unknown error, open '\\REMOTE_FOLDER_LAN_IP\NodeRealRobot\.vscode\settings.json')
>     at _.doReadAsFileStream (fileService.ts:457)
>     at errors.ts:26
> onUnexpectedError @ bootstrap-window.js:216
> bootstrap-window.js:213 [uncaught exception]: Error: Unable to read file '\\REMOTE_FOLDER_LAN_IP\NodeRealRobot\.vscode\tasks.json' (Unknown (FileSystemError): Error: UNKNOWN: unknown error, open '\\REMOTE_FOLDER_LAN_IP\NodeRealRobot\.vscode\tasks.json')
> 
> Error: Unable to read file '\\REMOTE_FOLDER_LAN_IP\NodeRealRobot\.vscode\tasks.json' (Unknown (FileSystemError): Error: UNKNOWN: unknown error, open '\\REMOTE_FOLDER_LAN_IP\NodeRealRobot\.vscode\tasks.json')
>     at _.doReadAsFileStream (file:///C:/Users/USER_NAME/AppData/Local/Programs/Microsoft VS Code/resources/app/out/vs/workbench/workbench.desktop.main.js:1667:417)
> onUnexpectedError @ bootstrap-window.js:213
> bootstrap-window.js:216 Error: Unable to read file '\\REMOTE_FOLDER_LAN_IP\NodeRealRobot\.vscode\tasks.json' (Unknown (FileSystemError): Error: UNKNOWN: unknown error, open '\\REMOTE_FOLDER_LAN_IP\NodeRealRobot\.vscode\tasks.json')
> 
> Error: Unable to read file '\\REMOTE_FOLDER_LAN_IP\NodeRealRobot\.vscode\tasks.json' (Unknown (FileSystemError): Error: UNKNOWN: unknown error, open '\\REMOTE_FOLDER_LAN_IP\NodeRealRobot\.vscode\tasks.json')
>     at _.doReadAsFileStream (fileService.ts:457)
>     at errors.ts:26
> onUnexpectedError @ bootstrap-window.js:216
> bootstrap-window.js:213 [uncaught exception]: Error: Unable to read file '\\REMOTE_FOLDER_LAN_IP\NodeRealRobot\.vscode\launch.json' (Unknown (FileSystemError): Error: UNKNOWN: unknown error, open '\\REMOTE_FOLDER_LAN_IP\NodeRealRobot\.vscode\launch.json')
> 
> Error: Unable to read file '\\REMOTE_FOLDER_LAN_IP\NodeRealRobot\.vscode\launch.json' (Unknown (FileSystemError): Error: UNKNOWN: unknown error, open '\\REMOTE_FOLDER_LAN_IP\NodeRealRobot\.vscode\launch.json')
>     at _.doReadAsFileStream (file:///C:/Users/USER_NAME/AppData/Local/Programs/Microsoft VS Code/resources/app/out/vs/workbench/workbench.desktop.main.js:1667:417)
> onUnexpectedError @ bootstrap-window.js:213
> bootstrap-window.js:216 Error: Unable to read file '\\REMOTE_FOLDER_LAN_IP\NodeRealRobot\.vscode\launch.json' (Unknown (FileSystemError): Error: UNKNOWN: unknown error, open '\\REMOTE_FOLDER_LAN_IP\NodeRealRobot\.vscode\launch.json')
> 
> Error: Unable to read file '\\REMOTE_FOLDER_LAN_IP\NodeRealRobot\.vscode\launch.json' (Unknown (FileSystemError): Error: UNKNOWN: unknown error, open '\\REMOTE_FOLDER_LAN_IP\NodeRealRobot\.vscode\launch.json')
>     at _.doReadAsFileStream (fileService.ts:457)
>     at errors.ts:26
> onUnexpectedError @ bootstrap-window.js:216
> console.ts:137 [IPC Library: File Watcher (nsfw)] Unhandled Promise Rejection:  Error: Path must be a valid path to a file or a directory.
>     at c:\Users\USER_NAME\AppData\Local\Programs\Microsoft VS Code\resources\app\node_modules.asar\vscode-nsfw\lib\src\index.js:79:11
> t.log @ console.ts:137

When i click on the remote folder after startup (exactly as when i take the JavaScript Profile) vscode don't show me this list of error, vscod show me this only after blocking  startup.

**Comment by bpasero** [comment]
Hm, not seeing anything in that profile for some reason. You should be able to start the CPU profile and then hit Ctrl+R to reload window and then stop the recording when done loading.

**Comment by FerrariAndrea** [comment]
Ok, now i get it. 
[CPU-20200318T185653.zip](https://github.com/microsoft/vscode/files/4350548/CPU-20200318T185653.zip)

At loading done i see this too in the console: 
> DevTools: CPU profile parser is fixing 48 missing samples.

And the same list of errors in the previous post.

(not i have a new folder on my workspace, i don't know if this is relevant)

**Comment by bpasero** [comment]
@FerrariAndrea yeah this time it looks better, I see a 30s grey area where nothing loads:

![image](https://user-images.githubusercontent.com/900690/76992553-b9451f00-694b-11ea-9878-057ccbd0d8fc.png)

Unfortunately I cannot tell what this is waiting for, this does not seem to be code we trigger, but something else. Sorry, I cannot be of more help I fear.

**Comment by FerrariAndrea** [comment]
It is not normal?  In your case this not happens?

Anyway this waiting time is not a big deal so thanks for your attention and i think that we can close this post. In case i will I'll find out new important info i will reopen it.

**Comment by bpasero** [comment]
> It is not normal? doesn't it happen in your case?

No, I do not think this is normal and I have never seen this, but I am also not having such a setup with a remote folder that is slow to be honest. 

Looking at the CPU profile again, one of the last things happening before the 30s delay seems to be resolving the configuration (setting) values from each folder. We do block the startup until we read all configuration I think, right @sandy081 ?

Would it be possible to time out after like 5s and then reconcile the configuration values later?

**Comment by sandy081** [comment]
> Would it be possible to time out after like 5s and then reconcile the configuration values later? Especially since the folder settings are probably not making a big impact on the overall experience right?

We cannot unblock everything but only workbench. ExtensionHost needs completely resolved configuration as extensions need initial configuration and accessing configuration is sync.

**Comment by bpasero** [comment]
@sandy081 yeah I am mainly concerned about the workbench UI that is not loading from that point on, I understand the extension host.

**Comment by sandy081** [comment]
I already follow this for remote folders / workspaces (in general for all with different FSP). Which FSP is this folder/workspace is using?

**Comment by bpasero** [comment]
@sandy081 this is the local one but with a slow behaving network share

**Comment by sandy081** [comment]
I would not go by time out. Instead have some user folder specific setting that will enable same pattern of loading folder settings as with other FSPs/

**Comment by bpasero** [comment]
Is that discoverable though, how would the user know?

**Comment by sandy081** [comment]
Good point, may be a notification asking user to delay the loading and reload. This can be stored in workspace state?

**Comment by bpasero** [comment]
@sandy081 just keep in mind that at that point no notification can show because the entire workbench is blocked. 

However, yes maybe this could work:
* we measure how slow config loading for each folder is
* for a folder with slow loading (e.g. >10s) we show a notification
* we store a setting in workspace state to skip this folder next time

However, I am unsure how the user could change this setting after the fact.

**Comment by FerrariAndrea** [comment]
I don't know if I am understanding and if I can be useful to you.
If whatever blocks the vscode startup is bind to the single folder, this is a good idea:

>- we measure how slow config loading for each folder is
>- for a folder with slow loading (e.g. >10s) we show a notification

and about this: 
> I am unsure how the user could change this setting after the fact.

Perhaps it is not necessary, the user will just click on the folders which he want work on.
Because seems that vscode behaves well after startup on the remote folders not loaded yet, as a previous post shows:
 ![image](https://user-images.githubusercontent.com/33545798/76577664-617b6380-64c6-11ea-95a8-91c6faa0d05c.png)

I don't know if you can block the loading of a folders that is takes too long time and then do this after on user click, as in the screen above (in this screen vscod is not blocked).

I didn't quite understand what the difference is between folders loader on startup of vscode and the loader on user click on the folder, in my mind it is the same code, so from my point of view it would be enough to postpone the loading of the folders to avoid the block, believing that the blocking occurs because something else is not yet ready when the folders are loaded at startup.

However, in my opinion, a folder that takes more than e.g. 10s certainly has problems and it is likely that the user knows it and therefore does not have to work on it. So if vscode needs block itself for loading something for a folder, it is preferable that vscode does this after an explicit user request.

I hope to be useful

**Comment by sandy081** [comment]
> However, I am unsure how the user could change this setting after the fact.

We can reset the state if it is faster?

**Comment by bpasero** [comment]
Yes, good idea. Since we always resolve the folder, even if it is slow, we can measure the time always and reset this state 👍

**Comment by ottopic** [comment]
I have the same problem #67934 #99975

**Comment by ottopic** [comment]
I'm using VS Code 1.46 on Mac OS 10.15.5 and my work's folder is on samba server.

```
/Volumes/space/libs/
/Volumes/space/commons/
/Volumes/space/Works/project1/
/Volumes/space/Works/project2/
```

I work on "space" volume but I don't need to index all folders because there are too many files.

I tried configuring Watcher Exclude and Search Exclude but indexing go on and causing traffic congestion on server.

```
"search.exclude": {
    "**/Volumes/space/commons/**": true,
    "**/Volumes/space/libs/**": true,
},
"files.watcherExclude": {
    "**/Volumes/space/commons/**": true,
    "**/Volumes/space/libs/**": true,
},
```

**Comment by davidavdav** [comment]
Hello,

We have a similar problem as @ottopic above https://github.com/microsoft/vscode/issues/92274#issuecomment-643258816 , we work with vscode local on macos, with files remote on a linux server with a mounted Synology NAS as filesystem, admittedly not the fastest of all remote working setups. 

Starting up vscode thrashes the filesystem completely, and there effectively is no language support (python) for the files we're editing, making it kind-of pointless to use the editor. 

Is there _any_ way to convince vscode to only index the files that are relevant, e.g., only python files, or the files actively open in the editor, or something?

Thanks
